### PR TITLE
Add restJson1 protocol test for content-type and POST operation with no input

### PR DESCRIPTION
**Pull request by haydenbaker**
#### Background
- Same as https://github.com/smithy-lang/smithy/pull/2322 but with a POST operation
- We should test both cases explicitly, since code-generation may have faulty assumptions (https://github.com/aws/aws-sdk-net/pull/3338)

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy [hunk 1]
```diff
@@ -291,20 +291,41 @@ structure TestNoPayloadInputOutput {
...
+apply TestPostNoInputNoPayload @httpRequestTests([
+    {
+        id: "RestJsonHttpPostWithNoInput",
+        documentation: "Serializes a POST request for an operation with no input, and therefore no modeled body",
+        protocol: restJson1,
+        method: "POST",
+        uri: "/no_input_no_payload",
+        body: "",
+        forbidHeaders: [
             "Content-Type"
```
Comment (gosar) [inline smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy:329]: looks like missed Content-Length here

File: smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy [hunk 2]
```diff
@@ -291,20 +291,41 @@ structure TestNoPayloadInputOutput {
...
 @http(uri: "/no_input_no_payload", method: "GET")
-operation TestNoInputNoPayload {
+operation TestGetNoInputNoPayload {
     output: TestNoPayloadInputOutput
 }
 
-apply TestNoInputNoPayload @httpRequestTests([
+@readonly
+@http(uri: "/no_input_no_payload", method: "POST")
+operation TestPostNoInputNoPayload {
```
Comment (gosar) [inline smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy:300]: I'd say add TestPostNoPayload (similar to TestNoPayload but with POST) also. That would actually more closely represent the issue represented in the .NET issue.

File: smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy (hunk 1 above)
Comment (haydenbaker) [reply smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy:329]: Clients SHOULD include `Content-Length: 0` on a POST without no body, hence the removal of `Content-Length` from POST: https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.2

File: smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy [hunk 3]
```diff
@@ -278,6 +278,53 @@ apply TestNoPayload @httpRequestTests([
     }
 ])
 
+/// This example operation serializes a request without an HTTP body for POST.
```
Comment (gosar) [inline smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy:281]:
```suggestion
/// This example POST operation serializes a request without a modeled HTTP body.
```

I'd say also change line 231 to
```
/// This example GET operation serializes a request without a modeled HTTP body.
```

File: smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy [hunk 4]
```diff
@@ -286,25 +333,46 @@ structure TestNoPayloadInputOutput {
 /// This example operation has no input and serializes a request without an HTTP body.
```
Comment (gosar) [inline smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy:333]:
```suggestion
/// This example GET operation has no input and serializes a request without an HTTP body.
```

and another comment for the POST operation.

File: smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy [hunk 5]
```diff
@@ -286,25 +333,46 @@ structure TestNoPayloadInputOutput {
...
 /// cannot produce an HTTP body.
 ///
 @readonly
 @http(uri: "/no_input_no_payload", method: "GET")
-operation TestNoInputNoPayload {
+operation TestGetNoInputNoPayload {
+    output: TestNoPayloadInputOutput
+}
+
+@readonly
```
Comment (gosar) [inline smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy:345]: I think it's better to put the operation and its tests in order.

i.e., 
/// docs about TestGetNoInputNoPayload
TestGetNoInputNoPayload
tests for TestGetNoInputNoPayload
/// docs about TestPostNoInputNoPayload
TestPostNoInputNoPayload
tests for TestPostNoInputNoPayload

File: smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy [hunk 6]
```diff
@@ -286,25 +333,46 @@ structure TestNoPayloadInputOutput {
 /// This example operation has no input and serializes a request without an HTTP body.
 ///
 /// These tests are to ensure we do not attach a body or related headers
-/// (Content-Length, Content-Type) to operations that semantically
+/// (Content-, Content-Type) to operations that semantically
```
Comment (gosar) [inline smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy:336]: it's confusing to have `Content-`

**Review by gosar — APPROVED** [review]

File: smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy [hunk 7]
```diff
@@ -278,33 +278,106 @@ apply TestNoPayload @httpRequestTests([
...
         forbidHeaders: [
-            "Content-Length",
+            "Content-Type",
+            "Content-Length"
+        ],
+        params: {}
+    }
+])
+
+/// This example POST operation has no input and serializes a request without a modeled HTTP body.
```
Comment (gosar) [inline smithy-aws-protocol-tests/model/restJson1/http-content-type.smithy:361]: nit: `modeled` is a bit odd here.

```suggestion
/// This example POST operation has no input and serializes a request without a HTTP body.
```

or just
```
/// This example POST operation has no input.
```
as the "These tests..." explains the rest.

(same for GET)

**Review by gosar — APPROVED** [review]

**Pull request merged**